WordPress/gutenberg
# Build: Reference Lodash as an external module

This pull request seeks to reference Lodash as an external module (window global), rather than include in the build of each bundle.

**Why?**

- Bundle size.
   - As implemented, we effectively include ~a copy of~ [a good chunk of](https://github.com/lodash/babel-plugin-lodash) Lodash in each bundle, upwards of a dozen or more copies in fact. There was some hope that the introduction of Webpack 4 and its handling of `sideEffects` would enable us to at least properly ["tree shake"](https://webpack.js.org/guides/tree-shaking/) to include only relevant Lodash modules. Preliminary testing, however, shows that `lodash-es` + `sideEffects` actually increases end bundle size. Through some debugging, it appears that with Lodash's internal dependency structure, all of Lodash is included anyways in most cases, with additional overhead of linking modules (even [concatenated](https://medium.com/webpack/brief-introduction-to-scope-hoisting-in-webpack-8435084c171f)). I think this warrants further exploration, though this has been time-consuming as-is, so may need to be deprioritized in favor of other pursuits.
- Build times.
   - Not measured, but since externals are effectively pointers to global objects, Webpack shouldn't need to perform module resolution and bundling internal to Lodash itself.

**Why not?** 

- Conflicts between Lodash and Underscore.js
   - This is not so much an issue with [Lodash's `_.noConflict`](https://lodash.com/docs/4.17.5#noConflict), though requires referencing Lodash on the global by its full name `lodash` (not the `_` shorthand).
- Conflicts between core's enqueue of Lodash vs. plugin enqueuing of Lodash.
   - This is perhaps more concerning, more-so with the no-conflict mode. Lodash is a very popular library, and it's quite likely plugins and themes are already enqueuing it. This becomes akin to themes enqueueing their own versions of jQuery, minus the forethought of dequeueing core's own. In any case, it's likely that one or the other of the two consumers, core or the plugin, would be working with a version they're not anticipating. Lodash has not seen catastrophic breaking changes for some time, so conflicts are not likely, though will be difficult to debug when they do occur.
   - We could enqueue with a handle more likely to be unique to core?

**Results** There is a total savings of 272.47kb of the total production (minified) bundle size, approximately 15.3% across all of Gutenberg's JavaScript bundles. Lodash's minified bundle is 71.4kb, making for a net savings of 201kb (11.2%). I believe this can be improved further with proper aliasing of `lodash-es` (see notes below).

Bundle|Before|After
---|---|---
`components/build/index.js`|349 KiB|305 KiB
`hooks/build/index.js`|10.1 KiB|10.1 KiB
`plugins/build/index.js`|40.4 KiB|20.9 KiB
`core-data/build/index`.js|54.5 KiB|34.6 KiB
`viewport/build/index.js`|34.3 KiB|12.2 KiB
`data/build/index.js`|74.7 KiB|50.4 KiB
`utils/build/index.js`|49.7 KiB|28 KiB
`i18n/build/index.js`|18.4 KiB|18.4 KiB
`element/build/index.js`|25 KiB|8.53 KiB
`editor/build/index.js`|280 KiB|233 KiB
`date/build/index.js`|186 KiB|186 KiB
`edit-post/build/index`.js|114 KiB|81.5 KiB
`blocks/build/index.js`|547 KiB|522 KiB

Before: 1783.1 KiB
After: 1510.63 KiB
Lodash: 71.4 KiB

Net savings: 201 KiB (11.2%)

**Implementation notes:**

_Babel Preset:_ We need to drop `babel-plugin-lodash` for this to work. It's not currently possible to extend our Babel preset configuration to omit this (even programmatically; I'm guessing it partially applies some behaviors upon being loaded?). I have inlined the Babel configuration into the Webpack configuration. This is obviously a temporary solution, and we should either drop this from the packaged Babel plugin, or otherwise allow to be configurable.

_`lodash-es`:_ Redux includes a couple references to `lodash-es/isPlainObject`. It would be ideal if these were consolidated to the external we have defined, as `lodash.isPlainObject`. I expected the proposed configuration would work and, in fact, the built bundle (before minification) includes inline comments about Redux' usage being for an "external" module, though it still includes the module inline. I have not found a related issue, though I wonder if this may be a bug within Webpack. In any case, this does not add much weight, but is duplicated nonetheless, and there is opportunity for further savings.

**Testing instructions:**

Verify there is no regressions in the build (including plugin-packed), and that the application loads and operates as expected.
